Change Log4j so the token is not logged to a file

pandaninjas opened the pull request:
This PR has been split into this and #619
Prevents Log4J from logging your token.

This PR prevents the export of logs, injects into log4j, and removes tokens from the log window, fixes #371 

Log4J versions implemented for:
- [x] 2.0-beta9
- [x] 2.17.1

Let me know if you have any concerns

pandaninjas committed:
Add logs to ignored export list, ensuring that logs (which may have session ID) don't get exported

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

flowln commented:
This may not work with instance types like those downloaded from CF, as they use `minecraft` instead of `.minecraft` at the moment.

Also, it doesn't seem to work at all the first time you try to export an instance. The issue here is that the function you modified only gets called if there's a `.packignore` file in the instance, and that only gets created after you did your first export.

pandaninjas committed:
Make prevention of log exports work for instances which don't use .minecraft or which don't use .packignore

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

Wyvest commented:
Not an active maintainer here or anything, but shouldn't session IDs just be filtered out of logs to prevent this from happening ever?

MicrocontrollersDev commented:
I agree with Wyvest here, most people that I have seen attempt to use this oversight to get a user's token don't use Prism's export instance feature but rather ask the user to upload crash reports directly from their .minecraft pretending to be part of a support team for a mod.

pandaninjas commented:
> I agree with Wyvest here, most people that I have seen attempt to use this oversight to get a user's token don't use Prism's export instance feature but rather ask the user to upload crash reports directly from their .minecraft pretending to be part of a support team for a mod.

I agree, what I have right now is a temporary mitigation, and I'll try to add filtering the Session ID from the logs soon

pandaninjas committed:
Remove Session ID from log window. However, logs aren't removed from log files like latest.log

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas commented:
The most recent commit filters out the Session ID line from the Minecraft Log window, but the log files in .minecraft like latest.log are written to by the Minecraft process, and filtering the Session ID out of those is much harder

DioEgizio commented:
But, is it filtered when uploading a log?

pandaninjas commented:
> But, is it filtered when uploading a log?

It definitely is

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

pandaninjas committed:
Add licenses for log4j

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Block Session IDs from being logged if the Minecraft version uses Log4J 2.0. However, this doesn't work on versions which don't use Log4J 2.0

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

Mqisty commented:
Let me know if I should close my issue

TheKodeToad commented:
Prism Launcher already seems to censor session IDs by replacing it with \<ACCESS TOKEN>. This is also reflected in the uploaded log. This actually works slightly better since it works on older versions.

Current behaviour:

![image](https://user-images.githubusercontent.com/57493648/201401841-ce5cdf89-c495-42f9-93ef-3f789c2bde72.png)



However, patching log4j to prevent it from even being in the files is a good new feature. I have seen malware (inside Minecraft plugins of all things) that has taken advantage of this.

DioEgizio commented:
> Prism Launcher already seems to censor session IDs by replacing it with <ACCESS TOKEN>. This is also reflected in the uploaded log. This actually works slightly better since it works on older versions.
> 
> Current behaviour:
> 
> ![image](https://user-images.githubusercontent.com/57493648/201401841-ce5cdf89-c495-42f9-93ef-3f789c2bde72.png)
> 
> However, patching log4j to prevent it from even being in the files is a good new feature. I have seen malware (inside Minecraft plugins of all things) that has taken advantage of this.

Not on very old versions

pandaninjas commented:
> > Prism Launcher already seems to censor session IDs by replacing it with . This is also reflected in the uploaded log. This actually works slightly better since it works on older versions.
> > Current behaviour:
> > ![image](https://user-images.githubusercontent.com/57493648/201401841-ce5cdf89-c495-42f9-93ef-3f789c2bde72.png)
> > However, patching log4j to prevent it from even being in the files is a good new feature. I have seen malware (inside Minecraft plugins of all things) that has taken advantage of this.
> 
> Not on very old versions

You also can find the Session ID even on the newest versions by going to your .minecraft/logs directory. Sometimes people get hacked from "developers" asking them to upload their log for "debugging"

TheKodeToad commented:
> > > Prism Launcher already seems to censor session IDs by replacing it with . This is also reflected in the uploaded log. This actually works slightly better since it works on older versions.
> > > Current behaviour:
> > > ![image](https://user-images.githubusercontent.com/57493648/201401841-ce5cdf89-c495-42f9-93ef-3f789c2bde72.png)
> > > However, patching log4j to prevent it from even being in the files is a good new feature. I have seen malware (inside Minecraft plugins of all things) that has taken advantage of this.
> > 
> > 
> > Not on very old versions
> 
> You also can find the Session ID even on the newest versions by going to your .minecraft/logs directory. Sometimes people get hacked from "developers" asking them to upload their log for "debugging"

The rest of this patch including the Log4j filtering is good.
I just though I'd point out that this code (as far as I can see) is redundant.

![image](https://user-images.githubusercontent.com/57493648/201467424-f8fd54ae-3e0d-4898-805e-565f5413c85c.png)

pandaninjas committed:
Remove code blocking Session ID in log windows

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

pandaninjas committed:
Remove Session ID in "Other logs" window

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Explicitly include QRegExp

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Fix C2665 by making the arguments explicit

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Use QRegularExpression, which is avaliable in both Qt5 and Qt6

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Remove #include <QtGlobal>

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Add Log4J 2.17 injection

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Change f_name to be more descriptive

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Add overrides to Log4jContextFactory

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Merge branch 'develop' into develop

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Readd deleted line

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Remove indentation

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

Scrumplex commented:
I am not sure if Prism is the right place to inject stuff into log4j. Generally, we should never modify behavior of the game or at the very least make it opt-out (or opt-in). Ideally, people would use mods to prevent this.

In general, I don't see much value in removing the session token from logfiles (the log4j injection), when an attacker could just read the `accounts.json` of the launcher instead. Session tokens expire fairly quickly compared to tokens in `accounts.json`

DioEgizio commented:
> I am not sure if Prism is the right place to inject stuff into log4j. Generally, we should never modify behavior of the game or at the very least make it opt-out (or opt-in). Ideally, people would use mods to prevent this.
> 
> In general, I don't see much value in removing the session token from logfiles (the log4j injection), when an attacker could just read the `accounts.json` of the launcher instead. Session tokens expire fairly quickly compared to tokens in `accounts.json`

Agreed, we should just remove it when uploading the log

pandaninjas commented:
> I am not sure if Prism is the right place to inject stuff into log4j. Generally, we should never modify behavior of the game or at the very least make it opt-out (or opt-in). Ideally, people would use mods to prevent this.
> 
> In general, I don't see much value in removing the session token from logfiles (the log4j injection), when an attacker could just read the `accounts.json` of the launcher instead. Session tokens expire fairly quickly compared to tokens in `accounts.json`

The motivation behind removing the session token from the logfiles isn't to prevent a malicious process from finding the session token. It's to prevent a common type of scam where someone pretending to be a developer of some sort asks a user to send their log files. I'll add an override to prevent this behavior.

DioEgizio commented:
again, just remove it while uploading the log. Injecting into log4j sounds very hacky and sketchy IMO

TheKodeToad commented:
I think manual uploading was also a concern.

pandaninjas committed:
Add an override

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4JInjector.java:
```diff
@@ -7,6 +7,9 @@ public class UniversalLog4JInjector {
     private static final Logger LOGGER = Logger.getLogger("UniversalLog4JInjector");
 
     public static void inject() {
+        if (System.getProperty("prism.log4j.inject").equals("false")) {
```
Maybe org.prismlauncher instead of prism?

TheKodeToad commented on libraries/launcher/org/apache/logging/log4j/LogManager.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * This class is a stub and is needed to compile the log4j injector
+ * Changing it may break things
+ **/
```
`**/` should be `*/`

TheKodeToad replied:
`**/` at the end is not conventional.

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4JInjector.java:
```diff
@@ -0,0 +1,41 @@
+    private static final Logger LOGGER = Logger.getLogger("UniversalLog4JInjector");
+
+    public static void inject() {
+        if (System.getProperty("prism.log4j.inject").equals("false")) {
```
This (I'm fairly sure) causes a NullPointerException. I also think "org.prismlauncher" would make more sense. While Boolean.getBoolean could be used, that would make it opt-in which would not be good.
```suggestion
        if ("false".equalsIgnoreCase(System.getProperty("org.prismlauncher.log4j.inject"))) {
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4JInjector.java:
```diff
@@ -0,0 +1,41 @@
+            l4jVersion = Class.forName("org.apache.logging.log4j.LogManager").getPackage().getImplementationVersion();
+            assert l4jVersion != null;
+        } catch (Exception e) {
+            LOGGER.log(Level.WARNING, "Log4J not found. Is this Minecraft using Log4J?");
```
This will create a warning on Minecraft versions that don't use Log4j which isn't great.

TheKodeToad commented on libraries/launcher/org/apache/logging/log4j/core/impl/Log4jContextFactory.java:
```diff
@@ -0,0 +1,35 @@
+
+        @Override
+        public LoggerContext getContext(String fqcn, ClassLoader loader, boolean currentContext) { 
+            throw new RuntimeException("This is a stub");
```
```suggestion
            throw new UnsupportedOperationException();
```
A message isn't necessary since this code won't be reached.

pandaninjas replied:
All of them are changed to UnsupportedOperationException

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/Log4JInjector20.java:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        @Override
+        public void info(String message) {
```
In newer versions of Minecraft it's logged with `debug`. However, it's unlikely the player will have it enabled. Still something to bear in mind.

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4JInjector.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.logging.Level;
+
+public class UniversalLog4JInjector {
+    private static final Logger LOGGER = Logger.getLogger("UniversalLog4JInjector");
```
In #291, java.util.logging was replaced with a custom Log class.

pandaninjas replied:
New commit uses the custom Log

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4JInjector.java:
```diff
@@ -0,0 +1,41 @@
+            try {
+                Log4JInjector2171.inject();
+            } catch (Exception e) {
+                LOGGER.log(Level.WARNING, "Unable to inject into log4j. Token may be written to a file as a result");
+                e.printStackTrace();
```
I feel that a constant would be good to avoid duplication.

pandaninjas replied:
Done

TheKodeToad reviewed (commented):
Sorry for the spam. I'd never used "Add single comment" and apparently it creates multiple reviews.

Just thought I'd suggest some changes. I didn't want to comment on every reference (since it's unimportant), but log4j and Log4J should be Log4j.

The code formatting is also inconsistent with the rest of the codebase, and if blocks should be converted to statements in some cases.

pandaninjas committed:
Make sure token isn't logged even if Minecraft uses .debug

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Replace .equals with null safe Objects.equals

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Apply changes from code review

Signed-off-by: pandaninjas <admin@malwarefight.gq>

pandaninjas committed:
Change RuntimeException to UnsupportedOperationException

Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4jInjector.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Objects;
+
+public class UniversalLog4jInjector {
+    private static final String EXCEPTION_DURING_INJECTION = "Unable to inject into Log4j. Token may be written to a file as a result";
```
Just a *tiny* suggestion.
```suggestion
    private static final String EXCEPTION_DURING_INJECTION = "Unable to inject into Log4j. Token may be written to a file as a result.";
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4jInjector.java:
```diff
@@ -0,0 +1,38 @@
+                e.printStackTrace();
+            }
+        } else {
+            Log.warning("Unable to inject into Log4j (Reason: Log4j version not supported). Token may be written to a file as a result");
```
```suggestion
            Log.warning("The active Log4j version is unsupported. Token may be written to a file as a result.");
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4jInjector.java:
```diff
@@ -0,0 +1,38 @@
+public class UniversalLog4jInjector {
+    private static final String EXCEPTION_DURING_INJECTION = "Unable to inject into Log4j. Token may be written to a file as a result";
+    public static void inject() {
+        if (Objects.equals(System.getProperty("org.prismlauncher.log4j.inject"), "false")) {
+            return;
+        }
```
Consistency
```suggestion
        if (Objects.equals(System.getProperty("org.prismlauncher.log4j.inject"), "false"))
            return;
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/UniversalLog4jInjector.java:
```diff
@@ -0,0 +1,38 @@
+        } catch (Exception e) {
+            return;
+        }
+        if (l4jVersion.startsWith("2.0")) {
```
I would remove the assert and do this instead. I'm not too familiar with this Java keyword, but I'm fairly sure it gets disabled. Additionally it would throw an exception which would be immediately ignored.
```suggestion
        if (l4jVersion == null)
            return;
 
        if (l4jVersion.startsWith("2.0")) {
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/Log4jInjector20.java:
```diff
@@ -0,0 +1,104 @@
+
+    public static void inject() {
+        try {
+            Field f = LogManager.class.getDeclaredField("factory");
+            f.setAccessible(true);
+            f.set(null, new WrappedLoggerContextFactory((Log4jContextFactory) LogManager.getFactory()));
```
```suggestion
            Field factory = LogManager.class.getDeclaredField("factory");
            factory.setAccessible(true);
            factory.set(null, new WrappedLoggerContextFactory((Log4jContextFactory) LogManager.getFactory()));
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/Log4jInjector20.java:
```diff
@@ -0,0 +1,104 @@
+
+        @Override
+        public void info(String message) {
+            if (message.startsWith("(Session ID is ")) {
+                return;
+            }
```
```suggestion
            if (message.startsWith("(Session ID is "))
                return;

```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/Log4jInjector20.java:
```diff
@@ -0,0 +1,104 @@
+            delegate.info(message);
+        }
+        
+        @Override
+        public void debug(String message) {
+            if (message.startsWith("(Session ID is ")) {
+                return;
+            }
+            delegate.info(message);
+        }
```
I think you meant to put this in Log4jInjector2171 - newer Minecraft versions use the debug log level.

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/Log4jInjector2171.java:
```diff
@@ -0,0 +1,95 @@
+        LogManager.setFactory(
+            new WrappedLoggerContextFactory((Log4jContextFactory) LogManager.getFactory())
+        );
+
```
Very small change.
```suggestion
```

TheKodeToad commented on libraries/launcher/org/prismlauncher/log4jinjector/Log4jInjector2171.java:
```diff
@@ -0,0 +1,95 @@
+
+        @Override
+        public void info(String message) {
+            if (message.startsWith("(Session ID is ")) {
+                return;
+            }
```
```suggestion
            if (message.startsWith("(Session ID is "))
                return;

```

TheKodeToad reviewed (changes requested):
Just saw a few things that you could change. I'd also suggest moving the note about the classes being stubs to package-info.java in org.apache.logging.

There are still some formatting inconsistencies.
The rest of classes are formatted as
```java
public class MyClass {
    // space
    private int a;
    private int b;

    void method();

    void otherMethod();
    // space
}
```
The rest of classes are also made final wherever possible.
I wouldn't suggest requesting another review though, since I'm just a contributor to this project.

pandaninjas committed:
Add code review change

Co-authored-by: TheKodeToad <freezingma1l@outlook.com>
Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Change Log4j not supported message

Co-authored-by: TheKodeToad <freezingma1l@outlook.com>
Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

pandaninjas committed:
Apply suggestions from code review

Co-authored-by: TheKodeToad <freezingma1l@outlook.com>
Signed-off-by: pandaninjas <101084582+pandaninjas@users.noreply.github.com>

TheKodeToad commented:
GitHub has leaked my "private" email address again :joy:. I find this ironic on a PR regarding security.

pandaninjas commented:
> GitHub has leaked my "private" email address again joy. I find this ironic on a PR regarding security.

You can turn on email privacy in your GitHub profile. If you would like, I can amend the commits and force push it

pandaninjas committed:
Apply changes from code review

Signed-off-by: pandaninjas <admin@malwarefight.gq>

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

TheKodeToad commented:
> > GitHub has leaked my "private" email address again joy. I find this ironic on a PR regarding security.
> 
> You can turn on email privacy in your GitHub profile. If you would like, I can amend the commits and force push it

I'm not really too bothered. I have 2fa.

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

TheKodeToad commented:
> Prism Launcher already seems to censor session IDs by replacing it with <ACCESS TOKEN>. This is also reflected in the uploaded log. This actually works slightly better since it works on older versions.
> 
> Current behaviour:
> 
> ![image](https://user-images.githubusercontent.com/57493648/201401841-ce5cdf89-c495-42f9-93ef-3f789c2bde72.png)
> 
> However, patching log4j to prevent it from even being in the files is a good new feature. I have seen malware (inside Minecraft plugins of all things) that has taken advantage of this.

Turns out filtering the log upload *is* required in Other Logs, just not the main log. mclo.gs doesn't filter it out (guess how I found out). Fortunately this PR *should* fix it, but please don't do this in the stable version.

pandaninjas commented:
> > Prism Launcher already seems to censor session IDs by replacing it with . This is also reflected in the uploaded log. This actually works slightly better since it works on older versions.
> > Current behaviour:
> > ![image](https://user-images.githubusercontent.com/57493648/201401841-ce5cdf89-c495-42f9-93ef-3f789c2bde72.png)
> > However, patching log4j to prevent it from even being in the files is a good new feature. I have seen malware (inside Minecraft plugins of all things) that has taken advantage of this.
> 
> Turns out filtering the log upload _is_ required in Other Logs, just not the main log. mclo.gs doesn't filter it out (guess how I found out). Fortunately this PR _should_ fix it, but please don't do this in the stable version.

Should I split this PR into log upload filtering and log4j injection?

Scrumplex commented:
> Should I split this PR into log upload filtering and log4j injection?

sounds like a good idea to me

pandaninjas committed:
Remove changes that are in #619

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Merge branch 'develop' of https://github.com/pandaninjas/PrismLauncher into develop

flowln commented on libraries/launcher/org/apache/logging/log4j/LogManager.java:
```diff
@@ -0,0 +1,15 @@
+    public static LoggerContextFactory getFactory() {
+        throw new UnsupportedOperationException();
+    }
+}
```
add newline pls

flowln commented on libraries/launcher/org/apache/logging/log4j/core/Logger.java:
```diff
@@ -0,0 +1,22 @@
+    public MessageFactory getMessageFactory() {
+        throw new UnsupportedOperationException();
+    }
+}
```
add newline here too pls

flowln commented on libraries/launcher/org/apache/logging/log4j/message/MessageFactory.java:
```diff
@@ -0,0 +1,7 @@
+ * This class is a stub and is needed to compile the log4j injector
+ * Changing it may break compilation
+ */
+public class MessageFactory {}
```
here as well!

flowln commented on libraries/launcher/org/apache/logging/log4j/spi/LoggerContext.java:
```diff
@@ -0,0 +1,7 @@
+ * This class is a stub and is needed to compile the log4j injector
+ * Changing it may break compilation
+ */
+public class LoggerContext {}
```
also here!

flowln commented on libraries/launcher/org/apache/logging/log4j/spi/LoggerContextFactory.java:
```diff
@@ -0,0 +1,19 @@
+        LoggerContext getContext(String fqcn, ClassLoader loader, Object externalContext, boolean currentContext, URI configLocation, String name);
+
+        void removeContext(org.apache.logging.log4j.spi.LoggerContext loggerContext);
+}
```
aaand here!

flowln commented on libraries/launcher/org/apache/logging/package-info.java:
```diff
@@ -0,0 +1,5 @@
+ * These classes are stubs that aid the Java compiler in compiling the Log4j injector
+ * Changing them may break compilation or cause Java to not be able to find certain methods
+ */
+package org.apache.logging;
```
and finally here

flowln reviewed (commented):
i don't know a whole lot of Java or meta, so excuse me if I make some mistake :sweat_smile: 
seems like the Log4j downloaded in my 1.19.3 instance is version 2.19.0, so not one of the ones you check for in the code. This is in meta [here](https://github.com/PrismLauncher/meta-launcher/blob/84dca764893453a78f75f19d3ef25741a45d2871/net.minecraft/1.19.3.json#L463).

I suppose, however, that there's mainly three types of users that are more likely to need this kind of change, due to surface area and stuff: People on 1.8.9, people on 1.12.2, and people on the newest possible version. The first two are already handled, while the latter is not. I suppose it'd make sense to also add support for the Log4j version used in the newest version, if that'd not me too much additional work?

pandaninjas committed:
Support 2.19.0

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Add newlines

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Merge branch 'PrismLauncher:develop' into develop

TheKodeToad commented:
The newest version doesn't log the access token by default. Edit: I just saw how little work it required so you may as well keep it.

flowln reviewed (approved):
thanks :)

TheKodeToad reviewed (approved):
Looks good! I don't have write access but I thought I should approve this because of my previous reviews.

Scrumplex commented:
I am kinda unsure here. This code seems fairly intrusive. Couldn't we just use log4j configurations instead?

This is an example config provided by Mojang to mitigate log4shell on some versions: https://launcher.mojang.com/v1/objects/02937d122c86ce73319ef9975b58896fc1b491d1/log4j2_112-116.xml

We could just take that and add a RegexFilter to `DENY` anything that matches the Session ID.

See https://logging.apache.org/log4j/2.x/manual/filters.html#RegexFilter

pandaninjas commented:
> I am kinda unsure here. This code seems fairly intrusive. Couldn't we just use log4j configurations instead?
> 
> This is an example config provided by Mojang to mitigate log4shell on some versions: https://launcher.mojang.com/v1/objects/02937d122c86ce73319ef9975b58896fc1b491d1/log4j2_112-116.xml
> 
> We could just take that and add a RegexFilter to `DENY` anything that matches the Session ID.
> 
> See https://logging.apache.org/log4j/2.x/manual/filters.html#RegexFilter

I will look into it.

pandaninjas committed:
Change to using log4j configuration files.

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Change to using log4j configuration files.

pandaninjas committed:
Merge branch 'develop' of https://github.com/pandaninjas/PrismLauncher into develop

pandaninjas committed:
DCO Remediation Commit for PandaNinjas <admin@malwarefight.gq>

I, PandaNinjas <admin@malwarefight.gq>, hereby add my Signed-off-by to this commit: 33a97b92d8a879f4511d8302a2eec9739d26e7b8

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas committed:
Remove import

Signed-off-by: PandaNinjas <admin@malwarefight.gq>

pandaninjas commented:
The most recent commits changes it to use a log4j config file

Scrumplex commented:
Recently, someone mentioned in our Discord or Matrix channels that regex-based filters might not be the best way to filter logs. They can be a major performance hit, especially when the game logs a lot of stuff due to errors or bugs.

I am not sure if there is a good way to implement this without affecting performance though

pandaninjas commented:
> Recently, someone mentioned in our Discord or Matrix channels that regex-based filters might not be the best way to filter logs. They can be a major performance hit, especially when the game logs a lot of stuff due to errors or bugs.
> 
> I am not sure if there is a good way to implement this without affecting performance though

There doesn't seem to be a way to do this without a regex filter.

DioEgizio commented:
i think because of the possible perf hit this should be an option disabled by default

DioEgizio commented:
Is this pr dead?

pandaninjas commented:
> Is this pr dead?

No, it's just rather difficult to make it toggle on. Right now how it works is that since log4j.xml is in the launch jar in the default package Log4j automatically finds it. I'm trying to find a solution right now.

TheKodeToad commented:
Maybe try doing something similar to #443 and have a separate jar that gets added if an option is on :3

TheKodeToad commented:
Actually maybe it's better to do this in meta?

pandaninjas commented:
> Actually maybe it's better to do this in meta?

It might be a good idea, but then we would have to patch jars before the meta serves it and it might break it for people using the meta for non-prism purposes

TheKodeToad commented:
You could use the special system variable (I forgot the name) to use a file config :sparkles:

TheKodeToad commented:
Closing this as the commit history has got pretty messy for such a small change - and it would probably need to be completely redone to make use of meta
